### PR TITLE
Allow students to complete an exercise

### PR DESCRIPTION
**Pull request by kntsoriano**
## Questions
1. I noticed in the API that there seems to be no code for publishing the completed solution. Will you be the one to add that @iHiD?
2. This also does not have the animations for the progress bar and modal. How do we plan to do this?

### CODE REVIEW

File: app/helpers/react_components/student/complete_exercise_button.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+        super(
+          "student-complete-exercise-button",
+          {
+            endpoint: Exercism::Routes.complete_api_solution_url(solution.uuid)
+          }
+        )
+      end
+
+      private
+      attr_reader :exercise
```
Comment (ErikSchierboom) [inline app/helpers/react_components/student/complete_exercise_button.rb]: Isn't this already created by the `initialize_with` functionality?

**Review by ErikSchierboom — APPROVED** [review]

**Comment by iHiD** [comment]
> 1. I noticed in the API that there seems to be no code for publishing the completed solution. Will you be the one to add that @iHiD?

Yep. This PR just needs to pass `publish: true` as part of the request to `/complete` if it should be published.

---

> 2. This also does not have the animations for the progress bar and modal. How do we plan to do this?

One to worry about later :)

File: app/helpers/react_components/student/complete_exercise_button.rb (hunk 1 above)
Comment (iHiD) [reply app/helpers/react_components/student/complete_exercise_button.rb]:
```suggestion
```

File: app/javascript/components/modals/exercise-completed-modal/ConceptIcon.tsx [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+  name,
+  size,
+}: {
+  name: string
+  size: 'small' | 'medium'
+}): JSX.Element => {
+  return (
+    <div className={`c-concept-icon c--${size}`}>{name.substring(0, 2)}</div>
+  )
+}
```
Comment (iHiD) [inline app/javascript/components/modals/exercise-completed-modal/ConceptIcon.tsx]: You can move this to `common`

File: app/javascript/components/modals/ExerciseCompletedModal.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+          <br />
+          {exercise.title}!
+        </div>
+      </header>
+      <div className="progressed-concepts">
+        {conceptProgressions.map((progression) => (
+          <ConceptProgression key={progression.name} {...progression} />
+        ))}
+      </div>
+      <div className="unlocks">
```
Comment (iHiD) [inline app/javascript/components/modals/ExerciseCompletedModal.tsx]: This whole `.unlocks` div shouldn't show if there are no unlocked exercises and no unlocked concepts.

File: app/javascript/components/modals/ExerciseCompletedModal.tsx [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+              {unlockedConcepts.length}{' '}
+              {pluralize('concept', unlockedConcepts.length)}
+            </h3>
+            {unlockedConcepts.map((concept) => (
+              <UnlockedConcept key={concept.name} {...concept} />
+            ))}
+          </div>
+        ) : null}
+      </div>
+      <button className="btn-cta">Continue</button>
```
Comment (iHiD) [inline app/javascript/components/modals/ExerciseCompletedModal.tsx]: This button should be a link to the exercise (`track_exercise_path(@track, @exercise)` in Rails terms).

**Review by iHiD — APPROVED** [review]
I've pushed a couple of small commits, and made a couple of small final suggestions. Then let's go!! :)

**Comment by iHiD** [comment]
Same intermittent failing tests here, @kntsoriano.

**Comment by iHiD** [comment]
Thanks @kntsoriano!

**Pull request merged**